### PR TITLE
'Days since last commit' metric in platform out of the box - SONAR-6353

### PR DESCRIPTION
**Pull request by teryk**
end to end test done: check the measure is correctly computed for a given project in the « Measures » page.

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputationSteps.java [hunk 1]
```diff
@@ -41,6 +41,7 @@
       ParseReportStep.class,
 
       // Persist data
+      ComputeAndPersistDaysSinceLastCommitStep.class,
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputationSteps.java]: I would rename the class "PersistDaysSinceLastCommitStep", no need to specify the "compute" word from my point of view

**Comment by GithubIntegration** [comment]
SonarQube analysis reported no new issue.

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStep.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+public class ComputeAndPersistDaysSinceLastCommitStep implements ComputationStep {
+
+  private static final long MILLISECONDS_PER_DAY = 1000 * 60 * 60 * 24L;
+
+  private final DbClient dbClient;
+  private final SourceLineIndex sourceLineIndex;
+  private final MetricCache metricCache;
+  private final System2 system;
+
+  private long lastCommitTimestamp = 0L;
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStep.java]: A step should be stateless, but having a shared variable is very dangerous if we forget to reset it. I would recommand to use it as a local variable and to use it as a method parameter.

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputationSteps.java (hunk 1 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputationSteps.java]: Compute is the "Decorator" part and persistence is the persistence part … for the time being we could consider 2 separate operations. WDYT ?

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStep.java (hunk 2 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStep.java]: Why the step should be stateless ? Each component are created for one analysis and one analysis only, executed sequentially ? Why do this step should be specifically stateless ?
In this specific case, I'm ok as being stateless is not too painful.

File: server/sonar-server/src/main/java/org/sonar/server/source/index/SourceLineIndex.java [hunk 3]
```diff
@@ -112,4 +113,20 @@ public SourceLineDoc getLine(String fileUuid, int line) {
     }
     throw new NotFoundException(String.format("No source found on line %s for file '%s'", line, fileUuid));
   }
+
+  public Date lastCommitedDateOnProject(String projectUuid) {
+    SearchRequestBuilder request = getClient().prepareSearch(SourceLineIndexDefinition.INDEX)
+      .setTypes(SourceLineIndexDefinition.TYPE)
+      .setSize(1)
+      .setQuery(QueryBuilders.boolQuery()
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/source/index/SourceLineIndex.java]: Specify the field to return with setFetchSource(new String[]{SourceLineIndexDefinition.FIELD_SCM_REVISION}) to decrease memory used by the query

File: server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStepTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    reportWriter.writeComponentScm(
+      BatchReport.Scm.newBuilder()
+        .setComponentRef(2)
+        .addChangeset(
+          BatchReport.Scm.Changeset.newBuilder()
+            .setDate(threeDaysAgo)
+        )
+        .build()
+      );
+    ComputationContext context = new ComputationContext(new BatchReportReader(dir), new ComponentDto().setId(456_789L).setUuid("project-uuid"));
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStepTest.java]: Setting the id is useless. I would also use ComponentTesting.newProject("project-uuid")

**Comment by julienlancelot** [comment]
Apply my feedback, then LGTM

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputationSteps.java (hunk 1 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputationSteps.java]: For the time being I'm not sure it makes sense to have these both verbs, but as you wish, we'll change that when working in the "Decorator" sprint.

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStep.java (hunk 2 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStep.java]: Indeed, I forget that the step is executed once per project. So ok to keep as it is.

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputationSteps.java (hunk 1 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputationSteps.java]: done

File: server/sonar-server/src/main/java/org/sonar/server/source/index/SourceLineIndex.java (hunk 3 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/source/index/SourceLineIndex.java]: done, I suppose you were talking about SourceLineIndexDefinition.FIELD_SCM_DATE

File: server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStepTest.java (hunk 4 above)
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStepTest.java]: done

File: server/sonar-server/src/main/java/org/sonar/server/source/index/SourceLineIndex.java (hunk 3 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/source/index/SourceLineIndex.java]: Indeed !

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStep.java (hunk 2 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStep.java]: Just one thing, if tomorrow we're using more than one worker, don't we gonna have an issue with this ?

Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputeAndPersistDaysSinceLastCommitStep.java]: Not if we load the components per analysis

**Comment by GithubIntegration** [comment]
SonarQube analysis reported no new issue.

**Pull request merged**